src/backend/api/Fusion.Resources.Domain/Queries/GetRelevantOrgUnits.cs

```diff
@@ -0,0 +1,332 @@
+
+                List<QueryRelevantOrgUnit> populatedOrgUnitResult = GetRelevantOrgUnits(orgUnits, orgUnitAccessReason);
+
+                var filteredOrgUnits = ApplyOdataFilters(request.Query, populatedOrgUnitResult.OrderBy(x => x.SapId));
+
+                var skip = request.Query.Skip.GetValueOrDefault(0);
+                var take = request.Query.Top.GetValueOrDefault(100);
+
+                filteredOrgUnits.DistinctBy(x => x.SapId);
+
+                var pagedQuery = QueryRangedList.FromEnumrableItems(filteredOrgUnits, skip, take);
+
+                return pagedQuery;
+            }
+
+            private static List<QueryRelevantOrgUnit> GetRelevantOrgUnits(IEnumerable<QueryRelevantOrgUnit> cachedOrgUnits, List<QueryOrgUnitReason> orgUnitAccessReason)
+            {
+                var endResult = new List<QueryRelevantOrgUnit>();
+                foreach (var org in orgUnitAccessReason)
+                {
+                    if (org?.FullDepartment != null && org?.Reason != null)
+                    {
+                        var alreadyInList = endResult.FirstOrDefault(x => x.FullDepartment == org.FullDepartment);
+
+                        if (alreadyInList is null)
+                        {
+                            var data = cachedOrgUnits.Where(x => x.FullDepartment == org?.FullDepartment.Replace('*', ' ').TrimEnd()).FirstOrDefault();
+
+                            if (data != null)
+                            {
+                                data.Reasons.Add(org.Reason);
+                                endResult.Add(data);
+                            }
+                        }
+                        else
+                        {
+                            if (!alreadyInList.Reasons.Contains(org.Reason))
+                            {
+                                alreadyInList.Reasons.Add(org.Reason);
+                            }
+                        }
+                    }
+                }
+
+                return endResult;
+            }
+
+            private static List<QueryRelevantOrgUnit> ApplyOdataFilters(ODataQueryParams filter, IEnumerable<QueryRelevantOrgUnit> orgUnits)
+            {
+
+                var filterGenerator = filter.GenerateFilters<QueryRelevantOrgUnit>(m =>
+                {
+                    m.SqlQueryMode = false;
+                    m.MapField("sapId", e => e.SapId);
+                    m.MapField("name", e => e.Name);
+                    m.MapField("shortName", e => e.ShortName);
+                    m.MapField("department", e => e.Department);
+                    m.MapField("fullDepartment", e => e.FullDepartment);
+                    m.MapField("reason", e => e.Reasons);
+                });
+                return orgUnits.Where(filterGenerator.FilterLambda.Compile()).ToList();
+            }
+            private async Task<QueryRelatedDepartments> ResolveCache(string fullDepartmentName, string cachekey, CancellationToken cancellationToken)
```

Comment:
scary.
first hit will cache based on fulldepartmentName argument.
second request will use items cached, but will never find anything.

Not sure if you need this at all. You at least need to refactor a bit and maybe add a proper name.